libvips/foreign/vips2jpeg.c

```diff
@@ -627,17 +628,24 @@ write_vips( Write *write, int qfac, const char *profile,
 	if( progressive ) 
 		jpeg_simple_progression( &write->cinfo ); 
 
-	/* Turn off chroma subsampling. Follow IM and do it automatically for
-	 * high Q. 
-	 */
-	if( no_subsample ||
-		qfac >= 90 ) { 
-		int i;
-
-		for( i = 0; i < in->Bands; i++ ) { 
-			write->cinfo.comp_info[i].h_samp_factor = 1;
-			write->cinfo.comp_info[i].v_samp_factor = 1;
-		}
+	switch (subsample_mode) {
+        case VIPS_FOREIGN_JPEG_SUBSAMPLE_ON:
+            break;
+        case VIPS_FOREIGN_JPEG_SUBSAMPLE_AUTO:
+            /* Turn off chroma subsampling. Follow IM and do it automatically for
+            * high Q.
+            */
+            if( qfac < 90 ) {
+                break;
+            }
+        case VIPS_FOREIGN_JPEG_SUBSAMPLE_OFF:
+        default: {
```

Comment:
Newline before `{`, like a func definition, and `for` formatted as:

```C
for( i = 0; i < xxx; i++ ) { 
```

Reply:
done